src/main/java/com/mongodb/hibernate/translate/attachment/Attachment.java

```diff
@@ -0,0 +1,110 @@
+ * @see AttachmentKey
+ */
+@SuppressWarnings("unchecked")
+public final class Attachment {
+
+    private @Nullable AttachmentKey<?> attachmentKey;
+    private @Nullable Object attachedValue;
+
+    @Override
+    public String toString() {
+        return "Attachment{" + "attachmentKey=" + attachmentKey + ", attachedValue=" + attachedValue + "}";
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.AccessModifier.PRIVATE)
+    boolean isBlank() {
+        return this.attachmentKey == null && this.attachedValue == null;
+    }
+
+    /**
+     * Detaches the attached value by invoking some visitor method which returns {@code void} but internally the method
+     * is expected to attach some value by invoking the other {@link #attach(AttachmentKey, Object)} method.
+     *
+     * @param attachmentKey type safe attachment type enum; detacher and attacher should share the identical
+     *     {@code AttachmentKey}; never null
+     * @param attacher some {@link Runnable} wrapper of visitor method which is supposed to attach some value
+     *     internally; never {@code null}
+     * @return the value attached by the attacher; after this method returns, the previous state will be restored; never
+     *     {@code null}
+     * @param <T> the expected Java generic type
+     */
+    public <T> T expect(AttachmentKey<T> attachmentKey, Runnable attacher) {
+        var previousAttachmentKey = this.attachmentKey;
+        try {
+            this.attachmentKey = attachmentKey;
+            this.attachedValue = null;
+            attacher.run();
+            if (this.attachedValue == null) {
+                throw new IllegalArgumentException(format("No value attached for attachment key: %s", attachmentKey));
+            }
+            return (T) this.attachedValue;
+        } finally {
+            // revert back old state
+            this.attachmentKey = previousAttachmentKey;
+            this.attachedValue = null;
+        }
+    }
+
+    /**
+     * Attaches value matching some {@link AttachmentKey}, so coordinating detacher could grab it by invoking the
+     * {@link #expect(AttachmentKey, Runnable)} method, sharing identical {@code AttachmentKey}.
+     *
+     * @param attachmentKey key for validation purpose; never null
+     * @param attachedValue value to be attached; never {@code null}
+     * @param <T> the expected Java generics type
+     */
+    public <T> void attach(AttachmentKey<T> attachmentKey, T attachedValue) {
+        if (attachmentKey != this.attachmentKey) {
+            throw new IllegalArgumentException(format(
+                    "Provided attachment key [%s] different from expected key [%s]",
+                    attachmentKey, this.attachmentKey));
+        }
+        if (this.attachedValue != null) {
+            throw new IllegalStateException(format(
+                    "Current attachment with key [%s] has attached value [%s] already when another new value is provided [%s]",
```

Comment:
If `Attachment` is not part of API, which I suspect is true, all the checks in `Attachment` are assertions. So if this is not part of API, let's make them assertions.

Reply:
It is definitively not part of API